File: scripts/audit_site.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "nokogiri"
require "open3"
require "uri"

root = File.expand_path(ARGV.fetch(0, "_site"))
errors = []
canonicals = {}
inline_scripts = []
documents = {}
site_host = "hecavex.com"

png_dimensions = lambda do |path|
  bytes = File.binread(path, 24)
  next nil unless bytes.start_with?("\x89PNG\r\n\x1A\n".b) && bytes.bytesize >= 24

  [bytes.byteslice(16, 4).unpack1("N"), bytes.byteslice(20, 4).unpack1("N")]
rescue Errno::ENOENT, Errno::EACCES
  nil
end

resolve_reference = lambda do |raw, html|
  value = raw.to_s.strip
  return { local: false } if value.empty? || value.start_with?("mailto:", "tel:", "data:", "javascript:", "//")

  if value.match?(%r{\Ahttps?://}i)
    uri = URI.parse(value)
    return { local: false } unless %w[http https].include?(uri.scheme.downcase) && uri.host&.downcase == site_host

    clean = uri.path.to_s
    fragment = uri.fragment
  elsif value.match?(%r{\A[a-z][a-z0-9+.-]*:}i)
    return { local: false }
  else
    clean = value.split(/[?#]/, 2).first
    fragment = value.include?("#") ? value.split("#", 2).last : nil
  end

  clean = URI::DEFAULT_PARSER.unescape(clean)
  target = if clean.empty?
             html
           elsif clean.start_with?("/")
             File.join(root, clean.delete_prefix("/"))
           else
             File.expand_path(clean, File.dirname(html))
           end
  candidates = [target]
  candidates << File.join(target, "index.html") if File.extname(target).empty? || target.end_with?("/")
  candidates << "#{target}.html" if File.extname(target).empty?
  { local: true, target: candidates.find { |candidate| File.file?(candidate) }, fragment: fragment }
rescue URI::InvalidURIError, ArgumentError
  { local: true, target: nil, fragment: nil }
end

Dir.glob(File.join(root, "**", "*.html")).sort.each do |html|
  relative = html.delete_prefix(root).tr("\\", "/")
  document = Nokogiri::HTML(File.read(html, encoding: "UTF-8"))
  documents[html] = document
  robots = document.at_css('meta[name="robots"]')&.[]("content").to_s.downcase
  indexable = !robots.include?("noindex")

  errors << "#{relative}: missing html lang" if document.at_css("html")&.[]("lang").to_s.strip.empty?
  errors << "#{relative}: missing title" if document.at_css("title")&.text.to_s.strip.empty?
  duplicate_ids = document.css("[id]").group_by { |node| node["id"] }.select { |id, nodes| !id.to_s.empty? && nodes.length > 1 }.keys
  errors << "#{relative}: duplicate ids #{duplicate_ids.join(', ')}" unless duplicate_ids.empty?

  %w[aria-labelledby aria-describedby].each do |attribute|
    document.css("[#{attribute}]").each do |node|
      node[attribute].to_s.split.each do |id|
        errors << "#{relative}: #{attribute} references missing id #{id}" unless document.at_xpath("//*[@id=#{id.inspect}]")
      end
    end
  end

  document.css("script:not([src])").each_with_index do |script, index|
    type = script["type"].to_s
    next unless type.empty? || type == "text/javascript"
    next if script.text.strip.empty?

    inline_scripts << { path: relative, index: index + 1, source: script.text }
  end
  document.css("script[src]").each do |script|
    source = script["src"].to_s
    if source.match?(%r{cdn\.jsdelivr\.net/npm/(?:mermaid|mathjax)|mathjax}i)
      errors << "#{relative}: prohibited diagram or mathematics runtime #{source}"
    end
  end

  outline = document.at_css("#article-outline")
  if outline
    outline_panel = outline.ancestors("section").find { |section| section["class"].to_s.split.include?("hx-toc") }
    errors << "#{relative}: article outline must be hidden until local JavaScript populates it" unless outline_panel&.key?("hidden")
  end

  research_link = document.at_css('.hx-workspace-switcher a[href="/en/research/"],.hx-workspace-switcher a[href="/lt/tyrimai/"]')
  if research_link
    research_route = relative.start_with?("/en/research/", "/lt/tyrimai/")
    marked_current = research_link["aria-current"] == "page"
    errors << "#{relative}: Research network link has incorrect aria-current state" unless research_route == marked_current
  end

  if indexable
    errors << "#{relative}: missing meta description" if document.at_css('meta[name="description"]')&.[]("content").to_s.strip.empty?
    canonical = document.at_css('link[rel="canonical"]')&.[]("href").to_s
    errors << "#{relative}: missing canonical" if canonical.empty?
    errors << "#{relative}: duplicate canonical also used by #{canonicals[canonical]}" if !canonical.empty? && canonicals[canonical]
    canonicals[canonical] = relative unless canonical.empty?
    %w[og:title og:description og:url og:image og:image:width og:image:height og:image:alt].each do |property|
      nodes = document.css("meta[property='#{property}']")
      errors << "#{relative}: expected one #{property}, found #{nodes.length}" unless nodes.length == 1
      errors << "#{relative}: empty #{property}" if nodes.length == 1 && nodes.first["content"].to_s.strip.empty?
    end
    %w[twitter:card twitter:title twitter:description twitter:image twitter:image:alt].each do |name|
      nodes = document.css("meta[name='#{name}'],meta[property='#{name}']")
      errors << "#{relative}: expected one #{name}, found #{nodes.length}" unless nodes.length == 1
      errors << "#{relative}: empty #{name}" if nodes.length == 1 && nodes.first["content"].to_s.strip.empty?
    end

    og_image = document.at_css('meta[property="og:image"]')&.[]("content").to_s
    twitter_image = document.at_css('meta[name="twitter:image"],meta[property="twitter:image"]')&.[]("content").to_s
    errors << "#{relative}: Open Graph and Twitter images differ" unless og_image.empty? || og_image == twitter_image
    errors << "#{relative}: social image must be a PNG" unless URI.parse(og_image).path.to_s.downcase.end_with?(".png")
    errors << "#{relative}: og:image:width must be 1200" unless document.at_css('meta[property="og:image:width"]')&.[]("content") == "1200"
    errors << "#{relative}: og:image:height must be 630" unless document.at_css('meta[property="og:image:height"]')&.[]("content") == "630"
    begin
      preview_uri = URI.parse(og_image)
      if preview_uri.host&.downcase == site_host
        preview_path = File.join(root, URI::DEFAULT_PARSER.unescape(preview_uri.path).delete_prefix("/"))
        dimensions = png_dimensions.call(preview_path)
        errors << "#{relative}: social image is missing or not a valid PNG (#{preview_uri.path})" unless dimensions
        errors << "#{relative}: social image must be 1200x630, found #{dimensions&.join('x')}" if dimensions && dimensions != [1200, 630]
      else
        errors << "#{relative}: social image must be hosted on #{site_host}"
      end
    rescue URI::InvalidURIError
      errors << "#{relative}: invalid social image URL #{og_image}"
    end
    scripts = document.css('script[type="application/ld+json"]')
    errors << "#{relative}: expected one consolidated JSON-LD graph, found #{scripts.length}" unless scripts.length == 1
    scripts.each do |script|
      begin
        data = JSON.parse(script.text)
        graph = data["@graph"]
        errors << "#{relative}: JSON-LD is not an @graph" unless graph.is_a?(Array)
        ids = Array(graph).filter_map { |node| node["@id"] }
        expected_ids = %w[
          https://hecavex.com/#organization
          https://hecavex.com/#deividas-lis
          https://hecavex.com/#website
          https://apt.hecavex.com/#website
          https://labs.hecavex.com/#website
          https://radar.hecavex.com/#website
        ]
        missing_ids = expected_ids - ids
        errors << "#{relative}: graph is missing shared HECAVEX identities: #{missing_ids.join(', ')}" unless missing_ids.empty?
      rescue JSON::ParserError => error
        errors << "#{relative}: invalid JSON-LD (#{error.message.lines.first.strip})"
      end
    end
    errors << "#{relative}: expected exactly one main landmark" unless document.css("main").length == 1
    errors << "#{relative}: missing h1" if document.css("h1").empty?
  end

  document.css("img").each { |image| errors << "#{relative}: image missing alt attribute" unless image.key?("alt") }
  document.css("input,select,textarea").each do |control|
    id = control["id"].to_s
    labelled = control["aria-label"] || control["aria-labelledby"] || control.ancestors("label").any? || (!id.empty? && document.at_css("label[for='#{id}']"))
    errors << "#{relative}: form control #{id.empty? ? control.name : id} has no accessible label" unless labelled
  end
  document.css("a[href],button,summary").each do |control|
    labelled_by = control["aria-labelledby"].to_s.split.filter_map { |id| document.at_xpath("//*[@id=#{id.inspect}]")&.text }.join(" ")
    visible_text = control.xpath('.//text()[not(ancestor::*[@aria-hidden="true"])]').text
    image_alt = control.css("img[alt]").map { |image| image["alt"] }.join(" ")
    accessible_name = [control["aria-label"], labelled_by, visible_text, image_alt, control["title"]].compact.join(" ").strip
    errors << "#{relative}: #{control.name} has no accessible name" if accessible_name.empty?
  end
  package = document.at_css(".hx-research-package")
  if package
    errors << "#{relative}: research package is missing its stable publication ID" if package.at_css(".hx-research-record .hx-mono")&.text.to_s.strip.empty?
    errors << "#{relative}: research package has fewer than eight standard metadata fields" if package.css(".hx-research-record > div").length < 8
    errors << "#{relative}: research package is missing visible update history" unless document.at_css(".hx-updates")
  end
  document.css("[href],[src]").each do |node|
    raw = node["href"] || node["src"]
    reference = resolve_reference.call(raw, html)
    next unless reference[:local]

    unless reference[:target]
      errors << "#{relative}: unresolved internal reference #{raw}"
      next
    end

    fragment = URI::DEFAULT_PARSER.unescape(reference[:fragment].to_s)
    next if fragment.empty? || File.extname(reference[:target]).downcase != ".html"

    target_document = documents[reference[:target]] ||= Nokogiri::HTML(File.read(reference[:target], encoding: "UTF-8"))
    anchors = target_document.css("[id],[name]").flat_map { |anchor| [anchor["id"], anchor["name"]] }.compact
    errors << "#{relative}: unresolved fragment #{raw}" unless anchors.include?(fragment)
  end
end

Dir.glob(File.join(root, "**", "*.xml")).sort.each do |xml|
  begin
    Nokogiri::XML(File.read(xml, encoding: "UTF-8")) { |config| config.strict }
  rescue Nokogiri::XML::SyntaxError => error
    relative = xml.delete_prefix(root).tr("\\", "/")
    errors << "#{relative}: invalid XML (#{error.message.lines.first.strip})"
  end
end

sitemap_path = File.join(root, "sitemap.xml")
if File.file?(sitemap_path)
  sitemap = File.read(sitemap_path, encoding: "UTF-8")
  %w[hecavex-media-kit-en.html hecavex-media-kit-lt.html].each do |filename|
    errors << "sitemap.xml: downloadable media asset is indexed (#{filename})" if sitemap.include?(filename)
  end
  %w[/en/projects/ /lt/projektai/ /en/glossary/ /lt/zodynas/].each do |public_path|
    errors << "sitemap.xml: missing public portfolio page #{public_path}" unless sitemap.include?("https://hecavex.com#{public_path}")
  end
else
  errors << "sitemap.xml: missing"
end

{
  "assets/media/hecavex-media-kit-en.html" => "https://hecavex.com/en/speaker/",
  "assets/media/hecavex-media-kit-lt.html" => "https://hecavex.com/lt/pranesejas/"
}.each do |path, expected_canonical|
  html = File.join(root, path)
  unless File.file?(html)
    errors << "#{path}: downloadable media kit missing"
    next
  end

  document = documents[html] ||= Nokogiri::HTML(File.read(html, encoding: "UTF-8"))
  robots = document.at_css('meta[name="robots"]')&.[]("content").to_s.downcase
  description = document.at_css('meta[name="description"]')&.[]("content").to_s.strip
  canonical = document.at_css('link[rel="canonical"]')&.[]("href").to_s
  errors << "#{path}: downloadable media kit must be noindex" unless robots.include?("noindex")
  errors << "#{path}: downloadable media kit is missing a description" if description.empty?
  errors << "#{path}: expected canonical #{expected_canonical}, found #{canonical}" unless canonical == expected_canonical
end

unless inline_scripts.empty?
  checker = <<~JAVASCRIPT
    const fs = require('node:fs');
    const scripts = JSON.parse(fs.readFileSync(0, 'utf8'));
    const invalid = [];
    for (const script of scripts) {
      try {
        new Function(script.source);
      } catch (error) {
        invalid.push({ path: script.path, index: script.index, message: error.message });
      }
    }
    process.stdout.write(JSON.stringify(invalid));
  JAVASCRIPT
  output, error, status = Open3.capture3("node", "-e", checker, stdin_data: JSON.generate(inline_scripts))
  if status.success?
    JSON.parse(output).each do |failure|
      errors << "#{failure['path']}: inline script #{failure['index']} has invalid JavaScript (#{failure['message']})"
    end
  else
    errors << "inline JavaScript audit could not run (#{error.lines.first.to_s.strip})"
  end
end

owned_stylesheet = File.join(root, "assets", "css", "hecavex.css")
if File.file?(owned_stylesheet)
  css = File.read(owned_stylesheet, encoding: "UTF-8")
  hx_classes = documents.values.flat_map do |document|
    document.css("[class]").flat_map { |node| node["class"].to_s.split.grep(/\Ahx-/) }
  end.uniq.sort
  structural_classes = %w[
    hx-author-proof hx-briefings-head hx-research-head hx-tag-index hx-topic-head hx-workspace-switcher
  ]
  unstyled = hx_classes.reject do |class_name|
    exact_selector = css.match?(/\.#{Regexp.escape(class_name)}(?![a-zA-Z0-9_-])/)
    base_class = class_name.split("--", 2).first
    styled_modifier = class_name.include?("--") && css.match?(/\.#{Regexp.escape(base_class)}(?![a-zA-Z0-9_-])/)
    exact_selector || styled_modifier || structural_classes.include?(class_name)
  end
  errors << "owned stylesheet has no selector for: #{unstyled.join(', ')}" unless unstyled.empty?
else
  errors << "owned stylesheet is missing"
end

if errors.any?
  warn errors.uniq.join("\n")
  exit 1
end

puts "Site audit passed: SEO, schema, social metadata, accessibility, links, fragments, XML, assets and owned component styles."

File: scripts/audit-site.mjs
#!/usr/bin/env node

import { readFile, readdir, stat } from 'node:fs/promises';
import { extname, join, relative, resolve } from 'node:path';

const root = resolve(process.argv[2] ?? 'dist');
const failures = [];
const analyticsToken = process.env.PUBLIC_HECAVEX_ANALYTICS_TOKEN?.trim() ?? '';
const analyticsSource = 'https://static.cloudflareinsights.com/beacon.min.js';

async function walk(directory) {
  const files = [];
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    const path = join(directory, entry.name);
    if (entry.isDirectory()) files.push(...await walk(path));
    else files.push(path);
  }
  return files;
}

async function isFile(path) { try { return (await stat(path)).isFile(); } catch { return false; } }
const files = await walk(root);
const htmlFiles = files.filter((file) => extname(file) === '.html');
const canonicalOwners = new Map();
const hreflangByCanonical = new Map();
let wideTableClasses = 0;
let promptClasses = 0;
let tableRegions = 0;
let evidenceFigures = 0;

function attr(tag, name) {
  return tag.match(new RegExp(`\\s${name}=(?:"([^"]*)"|'([^']*)')`, 'i'))?.slice(1).find((value) => value !== undefined) ?? '';
}

function stripMarkup(value) { return value.replace(/<script[\s\S]*?<\/script>/gi, '').replace(/<style[\s\S]*?<\/style>/gi, '').replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim(); }
function countOccurrences(value, needle) { return needle ? value.split(needle).length - 1 : 0; }

function resolveLocal(raw, currentFile) {
  if (!raw || /^(?:mailto:|tel:|data:|javascript:|#)/i.test(raw) || raw.startsWith('//')) return undefined;
  let url;
  try { url = new URL(raw, `https://hecavex.com/${relative(root, currentFile).replaceAll('\\', '/')}`); } catch { return { path: raw, target: '' }; }
  if (url.hostname !== 'hecavex.com') return undefined;
  const path = decodeURIComponent(url.pathname);
  const target = path.endsWith('/') ? join(root, path.slice(1), 'index.html') : join(root, path.slice(1));
  return { path, target };
}

for (const file of htmlFiles) {
  const route = `/${relative(root, file).replaceAll('\\', '/')}`;
  const html = await readFile(file, 'utf8');
  if (/\{:\s*[^}]+\}/i.test(html)) failures.push(`${route}: unprocessed legacy Markdown attribute marker`);
  if (/<p>\s*<img\b[^>]*>\s*<em>[\s\S]*?<\/em>\s*<\/p>/i.test(html)) failures.push(`${route}: evidence image and caption were not converted to a semantic figure`);
  wideTableClasses += (html.match(/class=["'][^"']*\bhx-table-wide\b/gi) ?? []).length;
  promptClasses += (html.match(/class=["'][^"']*\bprompt-(?:info|danger)\b/gi) ?? []).length;
  tableRegions += (html.match(/class=["'][^"']*\btable-scroll-region\b/gi) ?? []).length;
  evidenceFigures += (html.match(/class=["'][^"']*\bhx-evidence-figure\b/gi) ?? []).length;
  const shellDocument = !route.startsWith('/assets/media/');
  const articleDocument = /class=["'][^"']*\barticle-body\b[^"']*\bprose\b/i.test(html);
  const noindex = /<meta\s+[^>]*name=["']robots["'][^>]*content=["'][^"']*noindex/i.test(html);
  const expectedAnalyticsReferences = shellDocument && analyticsToken ? 1 : 0;
  const analyticsReferences = countOccurrences(html, analyticsSource);
  if (analyticsReferences !== expectedAnalyticsReferences) failures.push(`${route}: expected ${expectedAnalyticsReferences} Cloudflare Web Analytics beacon reference${expectedAnalyticsReferences === 1 ? '' : 's'}, found ${analyticsReferences}`);
  if (shellDocument && analyticsToken) {
    const tokenReferences = countOccurrences(html, analyticsToken);
    if (tokenReferences !== 1) failures.push(`${route}: expected one configured Cloudflare Web Analytics site-token reference, found ${tokenReferences}`);
    const beaconIndex = html.indexOf(analyticsSource);
    const bodyCloseIndex = html.toLowerCase().lastIndexOf('</body>');
    if (beaconIndex < 0 || bodyCloseIndex < 0 || beaconIndex > bodyCloseIndex) failures.push(`${route}: analytics loader is not placed before the closing body tag`);
    if (!/beacon\.type\s*=\s*['"]module['"]/.test(html)) failures.push(`${route}: analytics beacon is not created as a module script`);
    if (!/doNotTrack/.test(html)) failures.push(`${route}: analytics loader is missing the Do Not Track gate`);
  }
  if (/\{%|\{\{\s*(?:site|page|post)\./i.test(html)) failures.push(`${route}: unrendered template output remains`);
  const htmlTag = html.match(/<html\b[^>]*>/i)?.[0] ?? '';
  if (!attr(htmlTag, 'lang')) failures.push(`${route}: missing html lang`);
  if (!/<title>[^<]+<\/title>/i.test(html)) failures.push(`${route}: missing title`);
  const redirect = /http-equiv=["']refresh/i.test(html);
  if (!noindex && !redirect) {
    if (!/<meta\s+[^>]*name=["']description["'][^>]*content=["'][^"']+/i.test(html)) failures.push(`${route}: missing meta description`);
    const canonical = html.match(/<link\s+[^>]*rel=["']canonical["'][^>]*href=["']([^"']+)/i)?.[1];
    if (!canonical) failures.push(`${route}: missing canonical`);
    else if (canonicalOwners.has(canonical)) failures.push(`${route}: duplicate canonical also used by ${canonicalOwners.get(canonical)}`);
    else canonicalOwners.set(canonical, route);
    const hreflangLinks = (html.match(/<link\b[^>]*>/gi) ?? [])
      .filter((tag) => attr(tag, 'rel').toLowerCase() === 'alternate' && attr(tag, 'hreflang'));
    const hreflangMap = new Map();
    for (const tag of hreflangLinks) {
      const language = attr(tag, 'hreflang').toLowerCase();
      const href = attr(tag, 'href');
      if (hreflangMap.has(language)) failures.push(`${route}: duplicate hreflang ${language}`);
      else hreflangMap.set(language, href);
    }
    if (!hreflangMap.has('x-default')) failures.push(`${route}: missing hreflang x-default`);
    const rootLanguageSelector = route === '/index.html' && canonical === 'https://hecavex.com/';
    if (!rootLanguageSelector && hreflangMap.has('en') && hreflangMap.has('lt') && hreflangMap.get('x-default') !== hreflangMap.get('en')) failures.push(`${route}: bilingual x-default must resolve to the English counterpart`);
    if (canonical) hreflangByCanonical.set(canonical, hreflangMap);
    for (const property of ['og:title', 'og:description', 'og:url', 'og:image', 'og:image:width', 'og:image:height', 'og:image:alt']) if (!new RegExp(`<meta\\s+[^>]*property=["']${property.replace(':', '\\:')}["'][^>]*content=["'][^"']+`, 'i').test(html)) failures.push(`${route}: missing ${property}`);
    for (const name of ['twitter:card', 'twitter:title', 'twitter:description', 'twitter:image', 'twitter:image:alt']) if (!new RegExp(`<meta\\s+[^>]*(?:name|property)=["']${name.replace(':', '\\:')}["'][^>]*content=["'][^"']+`, 'i').test(html)) failures.push(`${route}: missing ${name}`);
    if ((html.match(/<main\b/gi) ?? []).length !== 1) failures.push(`${route}: expected one main landmark`);
    if (!/<h1\b[^>]*>[\s\S]*?<\/h1>/i.test(html)) failures.push(`${route}: missing h1`);
    const graphSource = html.match(/<script\s+[^>]*type=["']application\/ld\+json["'][^>]*>([\s\S]*?)<\/script>/i)?.[1];
    if (!graphSource) failures.push(`${route}: missing JSON-LD graph`);
    else {
      try {
        const graph = JSON.parse(graphSource)['@graph'];
        const nodes = Array.isArray(graph) ? graph : [];
        const ids = new Set(nodes.map((node) => node['@id']).filter(Boolean));
        for (const id of ['https://hecavex.com/#organization', 'https://hecavex.com/#deividas-lis', 'https://hecavex.com/#website', 'https://apt.hecavex.com/#website', 'https://labs.hecavex.com/#website', 'https://radar.hecavex.com/#website']) if (!ids.has(id)) failures.push(`${route}: JSON-LD is missing ${id}`);
        const organization = nodes.find((node) => node['@id'] === 'https://hecavex.com/#organization');
        if (!organization?.logo?.url || !organization?.logo?.width || !organization?.logo?.height) failures.push(`${route}: Organization JSON-LD is missing a complete publisher logo`);
        if (articleDocument) {
          const article = nodes.find((node) => node['@type'] === 'Article');
          if (!article?.mainEntityOfPage) failures.push(`${route}: Article JSON-LD is missing mainEntityOfPage`);
          if (article?.isAccessibleForFree !== true) failures.push(`${route}: Article JSON-LD must declare isAccessibleForFree`);
          if (!article?.image?.url) failures.push(`${route}: Article JSON-LD is missing its image`);
          const breadcrumb = nodes.find((node) => node['@type'] === 'BreadcrumbList');
          if (!breadcrumb || !Array.isArray(breadcrumb.itemListElement) || breadcrumb.itemListElement.length < 2) failures.push(`${route}: Article JSON-LD is missing a complete BreadcrumbList`);
        }
      } catch (error) { failures.push(`${route}: invalid JSON-LD (${error.message})`); }
    }
  }
  if (!redirect && shellDocument) {
    if (!/<header\b[^>]*class=["'][^"']*site-header[^"']*["'][^>]*data-portfolio-shell=["']v1["']/i.test(html)) failures.push(`${route}: missing portfolio shell contract marker`);
    if (!/<link\s+[^>]*rel=["']icon["'][^>]*href=["']\/favicon\.svg["']/i.test(html)) failures.push(`${route}: missing shared SVG favicon`);
    if (!/<link\s+[^>]*rel=["']apple-touch-icon["'][^>]*href=["']\/apple-touch-icon\.png["']/i.test(html)) failures.push(`${route}: missing shared Apple touch icon`);
    if (!/<link\s+[^>]*rel=["']manifest["'][^>]*href=["']\/site\.webmanifest["']/i.test(html)) failures.push(`${route}: missing origin web manifest`);
    const network = [...html.matchAll(/<nav\s+class=["']portfolio-navigation["'][\s\S]*?<\/nav>/gi)][0]?.[0] ?? '';
    let previous = -1;
    for (const label of ['Research', 'Radar', 'APT Notes', 'Labs', 'Data']) {
      const index = network.indexOf(`>${label}<`);
      if (index < previous || index < 0) failures.push(`${route}: portfolio navigation order is incorrect at ${label}`);
      previous = index;
    }
  }
  for (const image of html.match(/<img\b[^>]*>/gi) ?? []) {
    if (!/\salt=(?:"[^"]*"|'[^']*')/i.test(image)) failures.push(`${route}: image missing alt attribute`);
    if (articleDocument && !/\bsrc=["']\/assets\/img\/brand\//i.test(image)) {
      if (!/\swidth=(?:["']|[^\s>]+)/i.test(image) || !/\sheight=(?:["']|[^\s>]+)/i.test(image)) failures.push(`${route}: article image missing intrinsic dimensions`);
      if (!/\sloading=(?:["']|[^\s>]+)/i.test(image)) failures.push(`${route}: article image missing loading policy`);
      if (!/\sdecoding=(?:["']|[^\s>]+)/i.test(image)) failures.push(`${route}: article image missing decoding policy`);
    }
  }
  for (const tag of html.match(/<(?:a|link|script|img)\b[^>]*(?:href|src)=(?:"[^"]*"|'[^']*')[^>]*>/gi) ?? []) {
    const raw = attr(tag, tag.toLowerCase().startsWith('<a') || tag.toLowerCase().startsWith('<link') ? 'href' : 'src');
    const local = resolveLocal(raw, file);
    if (local && !(await isFile(local.target))) {
      const indexTarget = `${local.target}${extname(local.target) ? '' : '/index.html'}`;
      if (!(await isFile(indexTarget))) failures.push(`${route}: unresolved internal reference ${raw}`);
    }
  }
  if (!stripMarkup(html)) failures.push(`${route}: empty document`);
}

if (wideTableClasses === 0) failures.push('legacy wide-table classes were not restored');
if (promptClasses === 0) failures.push('legacy prompt classes were not restored');
if (tableRegions === 0) failures.push('semantic table overflow regions were not generated');
if (evidenceFigures === 0) failures.push('semantic evidence figures were not generated');

for (const file of files.filter((path) => ['.xml', '.json', '.txt'].includes(extname(path)))) {
  const source = await readFile(file, 'utf8');
  if (/\{%|\{\{\s*(?:site|page|post)\./.test(source)) failures.push(`${relative(root, file)}: unrendered template syntax remains`);
  if (extname(file) === '.json') { try { JSON.parse(source); } catch (error) { failures.push(`${relative(root, file)}: invalid JSON (${error.message})`); } }
  if (extname(file) === '.xml' && !/^<\?xml[\s\S]*<(?:feed|urlset)\b/.test(source)) failures.push(`${relative(root, file)}: malformed XML document`);
}

for (const identityFile of ['favicon.svg', 'favicon.ico', 'apple-touch-icon.png', 'icon-192.png', 'icon-512.png', 'site.webmanifest']) {
  if (!(await isFile(join(root, identityFile)))) failures.push(`shared identity asset is missing: ${identityFile}`);
}

const sitemapSource = await readFile(join(root, 'sitemap.xml'), 'utf8');
for (const block of sitemapSource.match(/<url>[\s\S]*?<\/url>/g) ?? []) {
  const canonical = block.match(/<loc>([^<]+)<\/loc>/)?.[1];
  if (!canonical) continue;
  const htmlAlternates = hreflangByCanonical.get(canonical);
  if (!htmlAlternates) {
    failures.push(`sitemap canonical has no indexable HTML owner: ${canonical}`);
    continue;
  }
  const sitemapAlternates = new Map(
    [...block.matchAll(/<xhtml:link\s+[^>]*hreflang="([^"]+)"[^>]*href="([^"]+)"\s*\/>/g)]
      .map((match) => [match[1].toLowerCase(), match[2]])
  );
  for (const language of new Set([...sitemapAlternates.keys(), ...htmlAlternates.keys()])) {
    const expected = sitemapAlternates.get(language);
    const actual = htmlAlternates.get(language);
    if (expected !== actual) failures.push(`${canonical}: sitemap/HTML hreflang ${language} mismatch (${expected ?? 'missing'} vs ${actual ?? 'missing'})`);
  }
}

for (const lang of ['en', 'lt']) {
  const records = JSON.parse(await readFile(join(root, lang, 'search.json'), 'utf8'));
  if (!Array.isArray(records) || records.length === 0) failures.push(`${lang}/search.json: search index is empty`);
  for (const [index, record] of records.entries()) {
    if (!String(record.searchText ?? '').trim()) failures.push(`${lang}/search.json: record ${index} is missing precomputed searchText`);
    if (!Array.isArray(record.keywords)) failures.push(`${lang}/search.json: record ${index} is missing SEO keyword search fields`);
    if (Object.hasOwn(record, 'content')) failures.push(`${lang}/search.json: record ${index} still exposes the legacy unprocessed content field`);
  }
}

for (const feedPath of ['en/feed.xml', 'lt/feed.xml', 'en/briefings/feed.xml', 'lt/apzvalgos/feed.xml']) {
  const feed = await readFile(join(root, feedPath), 'utf8');
  const updates = [...feed.matchAll(/<updated>([^<]+)<\/updated>/g)].map((match) => Date.parse(match[1]));
  if (updates.length < 2 || updates.some((value) => !Number.isFinite(value))) failures.push(`${feedPath}: invalid or empty Atom update chronology`);
  else if (updates[0] !== Math.max(...updates.slice(1))) failures.push(`${feedPath}: feed-level updated timestamp is not the newest entry update`);
}

const css = await readFile(join(root, 'assets', 'css', 'hecavex.css'), 'utf8');
for (const [token, pattern] of [
  ['content width', /--content:\s*94rem/], ['network row', /--network-row:\s*4rem/], ['product row', /--product-row:\s*3\.25rem/],
  ['desktop header offset', /--header-offset:\s*7\.25rem/], ['mark size', /\.brand img\s*\{[^}]*width:\s*2\.25rem/s],
  ['product link height', /\.product-navigation a\s*\{[^}]*min-height:\s*3\.25rem/s], ['1160px collapse', /@media \(max-width:\s*1160px\)/],
  ['display heading', /font-size:\s*clamp\(2\.5rem,\s*5vw,\s*4rem\)/]
]) if (!pattern.test(css)) failures.push(`portfolio CSS contract missing ${token}`);

if (failures.length) {
  console.error([...new Set(failures)].join('\n'));
  process.exit(1);
}
console.log(`Site audit passed: ${htmlFiles.length} HTML documents; shell, SEO, schema, social metadata, assets and internal links are intact.`);
